### PR TITLE
[Functest stabilization]: Properly limit bandwidth with migration policies

### PR DESCRIPTION
**Pull request by iholder101**
**What this PR does / why we need it**:
This PR uses migration policies to limit bandwidth instead of using Kubevirt CR. This enhances tests isolation as they're not dependent on a global resource. In addition, it allows more tests to run in parallel.

One of the tests that now uses this method is test 8482 which is currently quarantined.

The logs (see example [1]) show that the migration already is finished when the test failed on not finding any Prometheus metrics. This sounds familiar - many tests used stress in the past to delay migration progress and prevent it from finishing immediately, and this issue was fixed for other tests by limiting migration bandwidth. The same approach is used here.

[1] https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/logs/periodic-kubevirt-e2e-k8s-1.24-sig-compute-migrations/1579489650844635136

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by iholder-redhat** [comment]
FYI @xpivarc 
Let's hope CI gods are with us :crossed_fingers:

File: tests/utils.go [hunk 1]
```diff
@@ -2475,11 +2484,21 @@ func GetPolicyMatchedToVmi(name string, vmi *v1.VirtualMachineInstance, namespac
...
+	if namespace != nil {
+		namespace.Labels = namespaceLabels
+	}
 
 	return policy
 }
 
+// GetPolicyMatchedToVmi mutates the given vmi parameter by adding labels to it. Therefore, it's recommended
+// to use this function before creating the vmi. Otherwise, its labels need to be updated.
+func GetPolicyMatchedToVmi(vmi *v1.VirtualMachineInstance) *migrationsv1.MigrationPolicy {
```
Comment (rmohr) [inline tests/utils.go:2500]: Let's not prefix it with `Get` if it mutates the VMI. How aboute `PreparePolicyAndVMI`? Would at least be slightly better.

File: tests/utils.go [hunk 2]
```diff
@@ -2475,11 +2486,89 @@ func GetPolicyMatchedToVmi(name string, vmi *v1.VirtualMachineInstance, namespac
...
+	return GetPolicyMatchedToVmiWithNsAndVmiLabels(vmi, nil, 1, 0)
+}
+
+// ApplyPolicyToVmi creates a new migration policy with the provided policySpec and updates / creates the VMI so it would
+// be matched to this policy using a single vmi-level label.
+func ApplyPolicyToVmi(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineInstance, policySpec migrationsv1.MigrationPolicySpec) (*v1.VirtualMachineInstance, *migrationsv1.MigrationPolicy) {
```
Comment (rmohr) [inline tests/utils.go:2506]: So what confuses me a little bit. Can't you just create in the tests in question the VMIs with the right label, but just not create the policy yet, being able to avoid all patch operations entirely by just creating and deleting the policy in time? That would probably simplify all this and would make most of this helper stuff unnecessary.

Would that be possible?

File: tests/utils.go (hunk 1 above)
Comment (iholder-redhat) [reply tests/utils.go]: Naming stuff is hard.. :)
Thanks, nice suggestion.

File: tests/utils.go (hunk 2 above)
Comment (iholder-redhat) [reply tests/utils.go]: I wanted to provide a way of limiting a running vmi with a certain bandwidth, but on a second thought I think you're right. The helper function became more complex than I intended with no real use case in the tests.

Changed :+1:

File: tests/migration_test.go [hunk 3]
```diff
@@ -258,6 +257,7 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
 	BeforeEach(func() {
 		virtClient, err = kubecli.GetKubevirtClient()
 		Expect(err).ToNot(HaveOccurred())
+		migrationBandwidthLimit = resource.MustParse("1Ki")
```
Comment (xpivarc) [inline tests/migration_test.go:260]: nit: this can be const

File: tests/migration_test.go [hunk 4]
```diff
@@ -520,35 +512,17 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
...
+		}, 100*time.Second, 1*time.Second).ShouldNot(BeEmpty(), "no metrics with kubevirt_migrate_vmi prefix are found")
 		Expect(metrics).To(HaveLen(len(lines)))
 
 		By("Checking the collected metrics")
 		keys := getKeysFromMetrics(metrics)
 		for _, key := range keys {
 			value := metrics[key]
-			fmt.Fprintf(GinkgoWriter, "metric value was %f\n", value)
-			Expect(value).To(BeNumerically(">=", float64(0.0)))
+			Expect(value).ToNot(BeZero(), "metric value is not expected to be zero")
```
Comment (xpivarc) [inline tests/migration_test.go]: Is this correct?

File: tests/migration_test.go [hunk 5]
```diff
@@ -3642,18 +3649,10 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
...
 
-			runStressTest(vmi, stressDefaultVMSize, stressDefaultTimeout)
-
 			// execute a migration, wait for finalized state
 			By("Starting the Migration")
 			migration := tests.NewRandomMigration(vmi.Name, vmi.Namespace)
-			migrationUID := runMigrationAndCollectMigrationMetrics(vmi, migration, 180)
-
-			// check VMI, confirm migration state
-			tests.ConfirmVMIPostMigration(virtClient, vmi, migrationUID)
```
Comment (xpivarc) [inline tests/migration_test.go:3708]: Why do we delete this? I am not sure I understand the concern here.

File: tests/migration_test.go [hunk 6]
```diff
@@ -3632,7 +3632,7 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
 
 	})
 
-	Context("[Serial][QUARANTINE][test_id:8482] Migration Metrics", func() {
+	Context("[Serial][test_id:8482] Migration Metrics", func() {
```
Comment (xpivarc) [inline tests/migration_test.go:3635]: This needs to be reverted before we merge.

File: tests/migration_test.go (hunk 4 above)
Comment (iholder-redhat) [reply tests/migration_test.go]: Well, in theory these values could be negative (since it's a `float` type).
In practice, these metrics should never be positive as they describe things like how many VMs are pending, how much data has been transferred, etc. The purpose of this check if to ensure the metrics are non-zero (or IOW, had been initialized and show actual data).

If you think it's better, I'll revert this.

File: tests/migration_test.go (hunk 5 above)
Comment (iholder-redhat) [reply tests/migration_test.go:3708]: As I wrote in the description of https://github.com/kubevirt/kubevirt/pull/8586/commits/a75a440616fc58ac0a5b08eebe607c2abc636422:

> It's hard to find the exact timeout / bandwidth limitation configs to
ensure the migration both ends on one hand, and is occuring for enough
time to collect metrics about it. Since the test is not focused on
validating the migration should end, this shouldn't be part of the test
(also, there are many other tests are focused on validating a migration
finish).
>
> In addition, Eventually clause is now changed to keep trying to get
valid metrics of some of them are with a zero-value, something that is
acceptable in Prometheus' context.

TLDR: this isn't what the test focuses on and aims to validate in the first place, and it's complicating the test and causing it to be much less stable and more flaky. There are many other tests that are focused on that goal.

File: tests/migration_test.go (hunk 6 above)
Comment (iholder-redhat) [reply tests/migration_test.go:3635]: Yup, that's right :+1: 
I just wanted to make sure the tests pass. I'll re-push the PR.

File: tests/migration_test.go (hunk 5 above)
Comment (xpivarc) [reply tests/migration_test.go:3708]: I understand that part but did we see this "waiting for migration end" part to be flaky?

Comment (iholder-redhat) [reply tests/migration_test.go:3708]: Yes. From local testing I see that the migration sometime ends and sometimes the timeout exceeds. Essentially this is a matter of tuning the timeout and bandwidth parameters which is outside the scope of this test IMHO. We have a lot of other tests that are focused on migration getting converged.

**Comment by iholder-redhat** [comment]
*ping*
PR is ready to review for a while now :)

**Comment by jean-edouard** [comment]
@xpivarc were your concerns addressed?
@iholder-redhat the code looks goot and I like the idea of using policies instead of the CR to allow more tests to run in parallel, but why isn't this PR switching any test from serial to parallel? (or did I miss something?)

**Comment by iholder-redhat** [comment]
> @iholder-redhat the code looks good and I like the idea of using policies instead of the CR to allow more tests to run in parallel, but why isn't this PR switching any test from serial to parallel? (or did I miss something?)

Yup, you're very right, thanks!
Now test 8482 is not `[Serial]` anymore. Note that the other tests about migration cancellation were already parallel, I've just refactored the migration policy functions to be clearer and easier to use. Ideally more follow-up PRs would continue with this approach and make more tests run in parallel.

File: tests/utils.go [hunk 7]
```diff
@@ -2499,7 +2508,33 @@ func GetPolicyMatchedToVmi(name string, vmi *v1.VirtualMachineInstance, namespac
...
+}
+
+func PreparePolicyAndVMIWithBandwidthLimitation(vmi *v1.VirtualMachineInstance, bandwidth resource.Quantity) *migrationsv1.MigrationPolicy {
+	policy := PreparePolicyAndVMI(vmi)
+	policy.Spec.BandwidthPerMigration = &bandwidth
+
+	return policy
+}
+
+func CreateMigrationPolicy(virtClient kubecli.KubevirtClient, policy *migrationsv1.MigrationPolicy) *migrationsv1.MigrationPolicy {
```
Comment (jean-edouard) [inline tests/utils.go:2533]: That's 4 Policy functions in a row in this overloaded file, can they maybe move to their own policy.go somewhere? That way we'd reduce the size of this file instead of increasing it...

**Review by jean-edouard — COMMENTED** [review]
Sorry for the double-review but I caught a couple more things...

File: tests/utils.go (hunk 7 above)
Comment (iholder-redhat) [reply tests/utils.go]: done

**Comment by iholder-redhat** [comment]
> Sorry for the double-review but I caught a couple more things...

No need to apologize!
Thanks for your review and important notes :)

**Review by fossedihelm — COMMENTED** [review]
Thanks @iholder10! One small nit (feel free to ignore it)
/lgtm
/hold @xpivarc I just want to make sure all your concerns are disappeared! Please unhold when you think we are ready, thanks!

File: pkg/virt-controller/watch/migration_test.go [hunk 8]
```diff
@@ -1504,15 +1511,14 @@ var _ = Describe("Migration watcher", func() {
...
 
-				policy := tests.GetPolicyMatchedToVmi("testpolicy", vmi, &namespace, 4, 3)
-				_, exists := policy.Spec.Selectors.VirtualMachineInstanceSelector[labelKey]
+				policy := tests.PreparePolicyAndVMIWithNsAndVmiLabels(vmi, &namespace, 4, 3)
+				_, exists := policy.Spec.Selectors.VirtualMachineInstanceSelector[fmt.Sprintf(labelKeyFmt, policy.Name)]
 				Expect(exists).To(BeTrue())
 
 				By("Changing one of the policy's labels to it won't match to VMI")
-				policy.Spec.Selectors.VirtualMachineInstanceSelector[labelKey] = labelValue + "XYZ"
+				policy.Spec.Selectors.VirtualMachineInstanceSelector[labelKeyFmt] = "XYZ"
```
[the remark this answers is not in the document]
Comment (iholder101) [reply pkg/virt-controller/watch/migration_test.go]: Yup, you're right :) thanks for catching that!

**Comment by iholder101** [comment]
By liking [this comment](https://github.com/kubevirt/kubevirt/pull/8586#issuecomment-1315450330) I assume that comments are addressed.
/hold cancel
